src/Middleware/DoublePassMiddlewareWrapper.php

```
@@ -0,0 +1,88 @@
+<?php
+/**
+ * @see       https://github.com/zendframework/zend-stratigility for the canonical source repository
+ * @copyright Copyright (c) 2017 Zend Technologies USA Inc. (https://www.zend.com)
+ * @license   https://github.com/zendframework/zend-stratigility/blob/master/LICENSE.md New BSD License
+ */
+
+namespace Zend\Stratigility\Middleware;
+
+use Interop\Http\Server\MiddlewareInterface;
+use Interop\Http\Server\RequestHandlerInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Zend\Diactoros\Response;
+use Zend\Stratigility\Exception;
+
+/**
+ * Decorate double-pass middleware as PSR-15 middleware.
+ *
+ * Decorates middleware with the following signature:
+ *
+ * <code>
+ * function (
+ *     ServerRequestInterface $request,
+ *     ResponseInterface $response,
+ *     callable $next
+ * ) : ResponseInterface
+ * </code>
+ *
+ * such that it will operate as PSR-15 middleware.
+ *
+ * Neither the arguments nor the return value need be typehinted; however, if
+ * the signature is incompatible, a PHP Error will likely be thrown.
+ */
+class DoublePassMiddlewareWrapper implements MiddlewareInterface
+{
+    /**
+     * @var callable
+     */
+    private $middleware;
+
+    /**
+     * @var ResponseInterface
+     */
+    private $responsePrototype;
+
+    /**
+     * @throws Exception\MissingResponsePrototypeException if no response
+     *     prototype is present, and zend-diactoros is not installed.
+     */
+    public function __construct(callable $middleware, ResponseInterface $responsePrototype = null)
+    {
+        $this->middleware = $middleware;
+
+        if (! $responsePrototype && ! class_exists(Response::class)) {
+            throw Exception\MissingResponsePrototypeException::create();
```

Not sure if I like this way of throwing exceptions, I understand this when we pass some arguments to the static method and based on these args we create some customized message, but here maybe we should use just `throw new Exception\MissingResponsePrototypeException(...)` ?

---

It simplifies the code creating the exception in this case because the exception name indicates what went wrong: missing response prototype. This allows me to ensure the message is consistent if we throw it elsewhere later, and also allows me to test the exception itself if desired.